bundles/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/scalarweb/ScalarWebDeviceManager.java

```diff
@@ -0,0 +1,267 @@
+     * @throws URISyntaxException if a URI syntax exception occurs
+     */
+    public ScalarWebDeviceManager(final URL baseUrl, final ScalarWebContext context)
+            throws IOException, DOMException, MalformedURLException, URISyntaxException {
+        this(baseUrl, "1.0", new HashSet<>(), context);
+    }
+
+    /**
+     * Private contructor to create a device manager from the parameters
+     *
+     * @param baseUrl a non-null base URL
+     * @param version a non-null, non-empty API version
+     * @param serviceProtocols a non-null, possibly empty list of protocols
+     * @param context a non-null context
+     * @throws IOException if an IOException occurs contacting the device
+     * @throws DOMException if a DOMException occurs processing the XML from the device
+     * @throws MalformedURLException if there is a malformed URL (in the device XML)
+     * @throws URISyntaxException if an URI exception occurs
+     */
+    private ScalarWebDeviceManager(final URL baseUrl, final String version, final Set<ServiceProtocol> serviceProtocols,
+            final ScalarWebContext context)
+            throws IOException, DOMException, MalformedURLException, URISyntaxException {
+        Objects.requireNonNull(baseUrl, "baseUrl cannot be null");
+        Validate.notEmpty(version, "version cannot be empty");
+        Objects.requireNonNull(serviceProtocols, "serviceProtocols cannot be null");
+        Objects.requireNonNull(context, "context cannot be null");
+
+        this.version = version;
+        this.baseUrl = baseUrl;
+
+        final Gson gson = GsonUtilities.getApiGson();
+
+        final SonyTransportFactory transportFactory = new SonyTransportFactory(baseUrl, gson,
+                context.getWebSocketClient(), context.getScheduler());
+
+        try (final SonyHttpTransport httpTransport = SonyTransportFactory.createHttpTransport(baseUrl,
+                ScalarWebService.GUIDE)) {
+            // Manually create the guide as it's used to get service protocols and supported methods
+            // Must use alternative supported api since SupportedApi requires a guide!
+            final SupportedApi guideApi = SupportedApi.getSupportApiAlternate(ScalarWebService.GUIDE, httpTransport,
+                    logger);
+            final ScalarWebService guide = new ScalarWebService(transportFactory,
+                    new ServiceProtocol(ScalarWebService.GUIDE, Collections.singleton(SonyTransportFactory.HTTP)),
+                    version, guideApi);
+
+            final ServiceProtocols sps = guide.execute(ScalarWebMethod.GETSERVICEPROTOCOLS).as(ServiceProtocols.class);
+            for (final ServiceProtocol serviceProtocol : sps.getServiceProtocols()) {
+                // remove the one from the device descriptor above (keyed by name)
+                // the add this one (which has protocol information)
+                serviceProtocols.remove(serviceProtocol);
+                serviceProtocols.add(serviceProtocol);
```

Comment:
I just reused the constructor set since it was handy - the logic definitely needs to be here so I'm taking this to mean that I shouldn't modify the constructor set, create my own and use it instead (that way it's all contained within the constructor).

The reason this is needed - the device descriptor (from UPNP descriptor) describes services and then I ask the device itself for what services are available.  Even though they both come from the same device - the union of the two give me a complete picture of what services are available